src/transformers/tokenization_transfo_xl.py

```diff
@@ -295,6 +297,19 @@ def _tokenize(self, line, add_eos=False, add_double_eos=False):
         else:
             return symbols
 
+    def prepare_for_tokenization(self, text, **kwargs):
+
+        look_ahead_for_special_token = '(?=[{}])'.format(self.punctuation_symbols)
+        look_ahead_to_match_all_except_space = '(?=[^\s])'  # noqa: W605
+        punction_without_space_before = '[^\s][{}]'.format(self.punctuation_symbols)  # noqa: W605
+        # add spaces before punctuation symbols as should be done in transfo-xl
+        if "add_space_before_punct_symbol" in kwargs and kwargs["add_space_before_punct_symbol"]:
+            text = re.sub(r'' + look_ahead_for_special_token + look_ahead_to_match_all_except_space, r' ', text)
+        elif re.search(punction_without_space_before, text):
```

Comment:
If the `elif` statement would slow the tokenizer down too much for long texts that are "correctly" preprocessed, another possibility would be to simply add a warning every time transfo-xl tokenizier is used. What do you think?
@mfuntowicz , @thomwolf

Reply:
I would not put to much concern in the speediness of Python's tokenizers, they inherently slow. Also, TransfoXL already has it's fast counterpart, so it might mitigate the need for Python's one to be fast.

@patrickvonplaten: Did you observe the same issue with `TranformerXlTokenizerFast` ? If the issue is also present in the later, we need to have a look quickly on Rust side, as it'll be picked by default by AutoTokenizer.

Reply:
Just a thought: you are now creating the three regexes again and again on each call to prepare_for_tokenization(). (Compiling) regex is slow (or at least that's what I remember from a long time ago when I was switched Perl code to Python). Might it be an option to compile the regex into a regex object as a class instance attribute? That way the regex pattern just needs to be compiled once (in init).

Reply:
@BramVanroy - thanks! Just changed it to compilation at instantiation time

Reply:
@mfuntowicz 
I observe the same issue when using `TranformerXlTokenizerFast`.